src/Toml/Type/AnyValue.hs

```diff
@@ -82,8 +83,14 @@ reifyAnyValues _ []                 = Right []
 reifyAnyValues v (AnyValue av : xs) = sameValue v av >>= \Refl -> (av :) <$> reifyAnyValues v xs
 
 -- | Function for creating 'Array' from list of 'AnyValue'.
-toMArray :: [AnyValue] -> Maybe (Value 'TArray)
-toMArray [] = Just $ Array []
+toMArray :: [AnyValue] -> Either BiMapError (Value 'TArray)
+toMArray [] = Right $ Array []
 toMArray (AnyValue x : xs) = case reifyAnyValues x xs of
-    Left _     -> Nothing
-    Right vals -> Just $ Array (x : vals)
+    Left _     -> Left BiMapError
+    Right vals -> Right $ Array (x : vals)
+
+----------------------------------------------------------------------------
+-- BiMap error type
+----------------------------------------------------------------------------
+
+data BiMapError = BiMapError
```

Comment:
But all these function should return `Maybe, not `Either`, because if they fail, we know why, it's because of a type mismatch, we can use that knowledge in `BiMap.hs`.

Reply:
But if `BiMap` are `Either` implemented then it ask all functions used `BiMap` to be `Either` or to add convertor from `Maybe` to `Either`.

Reply:
Look at how `matchText` is implemented:
```haskell
matchText :: Value t -> Either BiMapError Text
matchText (Text s) = Right s
matchText _        = Left BiMapError
```
It can either succeed or fail, but it doesn't even know why it failed because we are pattern matching with `_`. A junction that fails without knowing why should always return `Nothing`.  Then in `BiMap.hs` you can do something fancy like
```
_Text :: BiMap BiMapError Text AnyValue
_Text = mkAnyValueBiMap 
    (\a@(AnyValue v) -> either (Left BiMapErrorTypeMistatch Text (valueType v)) Right $ matchText a) 
    Text
```

Reply:
Thank you.

Reply:
It is one-more-operation variant.
Proper is 
```
_Text :: BiMap BiMapError Text AnyValue
_Text = mkAnyValueBiMap
    (\a -> maybe (Left BiMapError) Right $ matchText a) Text
```

Reply:
Right. But maybe it's best to change `mkAnyValueBiMap` instead?

Reply:
OMG yes!